### PR TITLE
Support repofiles in precacher tool

### PR DESCRIPTION
**Pull request by dmcilvaney**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

The `precacher` tool will now also understand custom repo files passed (via `REPO_LIST=...`). For each repo file passed, the tool will query the associated repo and add the packages on that server to the URL list it uses (and outputs to `$(precache_state_dir)/repo_urls.txt`).

###### Change Log  

- `precacher` tool now supports `--repo-file` which can be used multiple times to add repo files to its list of package sources.

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

- Local testing

### CODE REVIEW

**Comment by dmcilvaney** [comment]
Requires https://github.com/microsoft/CBL-Mariner/pull/6216 before merging

File: toolkit/tools/precacher/precacher.go [hunk 1]
```diff
@@ -258,7 +290,34 @@ func getPackageRepoPaths(repoUrl string) (packages []string, err error) {
 	return
 }
 
-// downloadMissingPackages will attemp to download each package listed in rpmSnapshot that is not already present in the
+// getPackageRepoPathsFromUrl returns a list of packages available in the given repoUrl by running repoquery
```
Comment (cwize1) [inline toolkit/tools/precacher/precacher.go:293]: nit: `getPackageRepoUrlsFromRepoFile`

File: toolkit/tools/precacher/precacher.go [hunk 2]
```diff
@@ -141,31 +144,56 @@ func getAllRepoData(repoURLs []string, workerTar, buildDir, repoUrlsFile string)
...
+	for _, repoFile := range repoFiles {
+		var packageRepoUrls []string
 
-		// We will be searching by the name: "<name>-<version>.<distro>.<arch>", the results from the repoquery will be
-		// in the form of "<PARTIAL_URL>/<name>-<version>.<distro>.<arch>.rpm"
-		for _, packageRepoPath := range packageRepoPaths {
-			packageName := path.Base(packageRepoPath)
-			packageName = strings.TrimSuffix(packageName, ".rpm")
+		// Replace the existing repo file (if it exists) with the one that we want to query
+		destFile := path.Join(chrootRepoDir, path.Base(repoFile))
```
Comment (cwize1) [inline toolkit/tools/precacher/precacher.go:166]: Won't you get multiple repo files in the directory if each `repoFile` has a different file name?

File: toolkit/tools/precacher/precacher.go [hunk 3]
```diff
@@ -207,14 +235,18 @@ func createChroot(workerTar, chrootDir string, leaveChrootOnDisk bool) (queryChr
...
+	err = queryChroot.Run(func() error {
+		_, chrootErr := installutils.TdnfInstall(dnfUtilsPackageName, rootDir)
+		if chrootErr != nil {
+			chrootErr = fmt.Errorf("failed to install '%s':\n%w", dnfUtilsPackageName, err)
+			return chrootErr
 		}
-		return err
+
+		// Remove all existing repos, we will be adding the repo files we want to query later
+		chrootErr = os.RemoveAll("/etc/yum.repos.d")
```
Comment (cwize1) [inline toolkit/tools/precacher/precacher.go:248]: Does the installer chroot get reset each time a tool runs? (Doing this won't affect other tools, correct?)

Comment (dmcilvaney) [reply toolkit/tools/precacher/precacher.go:248]: Each time we create a chroot we get a fresh copy based on the `.tar.gz` file that contains the chroot.

File: toolkit/tools/precacher/precacher.go (hunk 2 above)
Comment (dmcilvaney) [reply toolkit/tools/precacher/precacher.go:166]: Ah, correct. Was a mixup of a few fixes. Originally it was always using the same destination file, so they would overwrite each other.

I've moved back to the old flow that places all the files, then calls the query command once.

**Comment by dmcilvaney** [comment]
Finally figured out why this tool was mangling my dev box, I had the flag backwards on the chroot generation. Just appended here since it's a quick fix. We should now see the expected error message if the previous invocation of the tool failed to clean up the chroot (which it does 100% of the time due to /dev/hugepages being busy when using repo files).

**Review by cwize1 — APPROVED** [review]

File: toolkit/tools/precacher/precacher.go [hunk 4]
```diff
@@ -258,7 +290,34 @@ func getPackageRepoPaths(repoUrl string) (packages []string, err error) {
...
 
-// downloadMissingPackages will attemp to download each package listed in rpmSnapshot that is not already present in the
+// getPackageRepoUrlsFromRepoFile returns a list of packages available in the given repoUrl by running repoquery
+func getPackageRepoUrlsFromRepoFile() (packageURLs []string, err error) {
+	const (
+		reqoqueryTool    = "repoquery"
+		randomNameLength = 10
+		printErrorOutput = true
+	)
+	// We have removed all other repo files from the chroot, so we can blindly enable all repos to get the full list of packages
```
Comment (PawelWMS) [inline toolkit/tools/precacher/precacher.go:300]: Where do we remove the other .repo files? I'm not sure where that logic is.

File: toolkit/tools/precacher/precacher.go [hunk 5]
```diff
@@ -258,7 +290,34 @@ func getPackageRepoPaths(repoUrl string) (packages []string, err error) {
...
+	var queryCommonArgList = []string{"-y", "-q", "--enablerepo=*", "-a", "--location"}
```
Comment (PawelWMS) [inline toolkit/tools/precacher/precacher.go:301]: Do we need `--enablerepo=*`? Won't `repoquery` use all repos by default?

File: toolkit/tools/precacher/precacher.go [hunk 6]
```diff
@@ -258,7 +290,34 @@ func getPackageRepoPaths(repoUrl string) (packages []string, err error) {
 	return
 }
 
-// downloadMissingPackages will attemp to download each package listed in rpmSnapshot that is not already present in the
+// getPackageRepoUrlsFromRepoFile returns a list of packages available in the given repoUrl by running repoquery
+func getPackageRepoUrlsFromRepoFile() (packageURLs []string, err error) {
```
Comment (PawelWMS) [inline toolkit/tools/precacher/precacher.go:294]: Nit-pick:
```suggestion
// getPackageRepoUrlsFromRepoFiles returns a list of packages available in all RPM repos listed in the system's .repo files.
func getPackageRepoUrlsFromRepoFiles() (packageURLs []string, err error) {
```

File: toolkit/tools/precacher/precacher.go (hunk 4 above)
Comment (PawelWMS) [reply toolkit/tools/precacher/precacher.go:300]: Found it!

**Review by PawelWMS — APPROVED** [review]
Left just a couple of small change requests.

File: toolkit/tools/precacher/precacher.go (hunk 5 above)
Comment (dmcilvaney) [reply toolkit/tools/precacher/precacher.go:301]: I think `enabled=0` in the repofile will disable it by default. I'm assuming that you want to use all the repos you pass in here.

File: toolkit/tools/precacher/precacher.go (hunk 6 above)
Comment (dmcilvaney) [reply toolkit/tools/precacher/precacher.go:294]: Ah, hold over from when it did them one at a time.

**Pull request merged**